src/backends/wasm_js.rs

```diff
@@ -7,32 +7,48 @@ pub use crate::util::{inner_u32, inner_u64};
 #[cfg(not(all(target_arch = "wasm32", any(target_os = "unknown", target_os = "none"))))]
 compile_error!("`wasm_js` backend can be enabled only for OS-less WASM targets!");
 
-use js_sys::{global, Uint8Array};
-use wasm_bindgen::{prelude::wasm_bindgen, JsCast, JsValue};
+#[cfg(target_feature = "atomics")]
+use js_sys::Uint8Array;
+use wasm_bindgen::{prelude::wasm_bindgen, JsValue};
 
-// Size of our temporary Uint8Array buffer used with WebCrypto methods
-// Maximum is 65536 bytes see https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues
-const CRYPTO_BUFFER_SIZE: u16 = 256;
+// Maximum buffer size allowed in `Crypto.getRandomValuesSize` is 65536 bytes.
+// See https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues
+const MAX_BUFFER_SIZE: u16 = 256;
```

Comment:
This is still an artifact of when we used to cache the `Uint8Array`. I assume this was lowered from 65536 so we don't hog 64kB in the background even when that's not needed.

For the happy path, non-atomic, this is pointless ofc because we don't actually create any buffer.

For atomic, the question is if it actually makes sense to create a small buffer and call `getRandomValues()` more often or is it better to create the maximum bytes `getRandomValues()` in favor of reducing the amount of calls.

So I did some local benchmarking (I apologize for the formatting):

| bytes |   base   |       64        |       128       |       256       |       512       |      1024       |      2048       |      4096       |      8192       |      16384      |      32768      |      65536      |
| ----: | -------: | --------------: | --------------: | --------------: | --------------: | --------------: | --------------: | --------------: | --------------: | --------------: | --------------: | --------------: |
|     8 |    435ns |   1.152µs  265% |                 |                 |                 |                 |                 |                 |                 |                 |                 |                 |
|    16 |    482ns |   1.229µs  255% |                 |                 |                 |                 |                 |                 |                 |                 |                 |                 |
|    32 |    555ns |   1.297µs  234% |                 |                 |                 |                 |                 |                 |                 |                 |                 |                 |
|    64 |    601ns |   1.381µs  230% |                 |                 |                 |                 |                 |                 |                 |                 |                 |                 |
|   128 |    687ns |   2.378µs  346% |   1.483µs  216% |                 |                 |                 |                 |                 |                 |                 |                 |                 |
|   256 |    847ns |   4.274µs  505% |   2.527µs  298% |   1.686µs  199% |                 |                 |                 |                 |                 |                 |                 |                 |
|   512 |  1.202µs |   7.994µs  665% |   4.618µs  384% |   2.944µs  245% |   2.171µs  181% |                 |                 |                 |                 |                 |                 |                 |
|  1024 |  1.891µs |  15.772µs  834% |   8.797µs  465% |   5.412µs  286% |   3.824µs  202% |   3.112µs  165% |                 |                 |                 |                 |                 |                 |
|  2048 |   3.28µs |  30.857µs  941% |  17.182µs  524% |   10.23µs  312% |   6.952µs  212% |   5.459µs  166% |   4.973µs  152% |                 |                 |                 |                 |                 |
|  4096 |   6.07µs |  59.789µs  985% |  32.748µs  540% |   19.68µs  324% |   13.07µs  215% |  10.094µs  166% |   8.944µs  147% |  10.097µs  166% |                 |                 |                 |                 |
|  8192 | 11.594µs | 117.864µs 1017% |  64.888µs  560% |  38.477µs  332% |  25.404µs  219% |  19.133µs  165% |   16.29µs  141% |  16.951µs  146% |  17.949µs  155% |                 |                 |                 |
| 16384 | 22.497µs | 236.436µs 1051% | 125.872µs  560% |   75.82µs  337% |  49.249µs  219% |  36.608µs  163% |  31.173µs  139% |  29.977µs  133% |  29.801µs  132% |  32.536µs  145% |                 |                 |
| 32768 | 45.626µs | 456.666µs 1001% | 251.232µs  551% | 151.411µs  332% |  98.836µs  217% |   72.35µs  159% |   59.47µs  130% |  55.756µs  122% |  54.609µs  120% |  56.186µs  123% |   63.44µs  139% |                 |
| 65536 | 92.087µs |  935.13µs 1015% |  512.67µs  557% |  301.09µs  327% | 196.186µs  213% |  144.02µs  156% | 118.095µs  128% |  105.63µs  115% |  102.41µs  111% |   104.2µs  113% |  110.35µs  120% | 118.917µs  129% |

"bytes": how many bytes we are trying to get
"base": baseline time, representing 100%, by using the non-atomic path (with the fixed maximum chunk size of 65536)
all the other columns: atomic path and how many bytes we use for the `Uint8Array` buffer

I skipped testing for cases where we are trying retrieve less bytes than the buffer size, as it should result in the same time.

The result seems clear: taking the maximum amount of bytes and reducing the calls to `getRandomValues()` seems to always yield the best result.

So my proposal is:
- Fix the chunk size for the happy path to be the maximum size: 65536.
- As we don't cache the `Uint8Array`, we should not create one with a fixed size, but `min(dest.len(), FIXED_SIZE)`.
- Increase the maximum size of the buffer in the atomic path to 65536 as well.

---

Let me know if you want me to apply this here or in a follow-up.

Reply:
I agree that we should use 64K chunk size for the non-atomic code path. But I think we should keep the 256 byte buffer for the atomic path. The main use-case for `getrandom` is generation of seeds and cryptographic material, so it's rarely used for generation of big random buffers. For example, in the [WASI p2 backend](https://github.com/rust-random/getrandom/blob/master/src/backends/wasi_p2.rs#L21-L25) we even use API which generates 64-bit integers instead of the buffer-based one.

Reply:
This is what I'm proposing to account for this:
```rust
    let buf_len = u32::min(
        dest.len().try_into().unwrap_or(MAX_BUFFER_SIZE.into()),
        MAX_BUFFER_SIZE.into(),
    );
    let buf = Uint8Array::new_with_length(buf_len);
```
So we will always use the minimum amount necessary. This is an improvement over the current state already, because right now we allocate 256 bytes even if we only need 8 bytes.

Reply:
Looks good to me!

Reply:
Done.